.github/workflows/release.yml

```diff
@@ -0,0 +1,96 @@
+name: release
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+
+jobs:
+  collect-modules:
+    runs-on: ubuntu-latest
+    outputs:
+      modules: ${{ steps.collect-modules.outputs.modules }}
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: collect modules
+        id: collect-modules
+        run: |
+          modules=$(find . -name .releaserc.yaml | \
+            xargs -I {} dirname {} | \
+            jq -R '[.]' | jq -s -c 'add')
+
+            echo "modules=$modules" >> $GITHUB_OUTPUT
+
+  release-modules:
+    needs: ["collect-modules"]
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        module: ${{fromJson(needs.collect-modules.outputs.modules)}}
+      fail-fast: false
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: cloudsmith-io/cloudsmith-cli-action@v1.0.3
+        with:
+          api-key: ${{ secrets.CLOUDSMITH_API_KEY }}
+
+      - name: Fetch tags
+        run: git fetch --prune --tags
+
+      - name: collect module info
+        id: module-info
+        run: |
+          module_path="${{ matrix.module }}"
+          module_name=$(yq .module.name $module_path/.releaserc.yaml)
+          module_version=$(yq .module.version $module_path/.releaserc.yaml)
+
+          tag_exists=$(git tag -l | grep -q $module_name-$module_version && echo true || echo false)
+
+          echo "module_name=$module_name" >> $GITHUB_OUTPUT
+          echo "module_version=$module_version" >> $GITHUB_OUTPUT
+          echo "tag_exists=$tag_exists" >> $GITHUB_OUTPUT
+
+      - name: publish module
+        if: steps.module-info.outputs.tag_exists == 'false'
+        run: |
+          module_path="${{ matrix.module }}"
```

Comment:
In general, I'd consider putting these scripts into `.sh` files in the repo, so that you could run them locally for testing, get better syntax highlighting and potentially `shellcheck` as well